### PR TITLE
Starting work on a deserializer

### PR DESCRIPTION
**Pull request by gnarf**

### CODE REVIEW

**Comment by evmiguel** [comment]
I like how BlockProperties extends SerializedBlockObject, and how makeIterable is cleaned up

File: src/sb3-serialize.ts [hunk 1]
```diff
@@ -0,0 +1,25 @@
+import { BlockProperties, SerializedBlock, SerializedBlocks, SerializedBlockObject } from './abstracts';
+
+const deserializedCache = new WeakMap<SerializedBlock, BlockProperties>();
+
+export function* deserializeBlocks(blocks: SerializedBlocks): Iterator<BlockProperties> {
```
Comment (mzgoddard) [inline src/sb3-serialize.ts]: We could generalize the generator parts into the generator folder like:

```js
function* iterValues (blocks) {
    for (const id in blocks) {
        yield blocks[id];
    }
}

function* iterCacheMap<T, U> (cache: WeakMap<T, U>, item: Iterator<T>, prediate: (T) => U) : Iterator<U> {
    for (const item of iter) {
        let result = cache.get(item);
        if (!result) {
            result = prediate(item);
            cache.set(item, result);
        }
        return result;
    }
}

function deserializeBlocks (blocks) {
    return iterCacheMap(deserializedCache, iterValues(blocks), serialized => {
        // ...
        return block;
    });
}
```

File: src/sb-util.ts [hunk 2]
```diff
@@ -178,16 +179,7 @@ export class Sprite extends SpriteCollection {
...
-        // eslint-disable-next-line @typescript-eslint/no-explicit-any
-        const blocksObj: Record<string, any> = this.prop('blocks');
-        const allBlocks: Iterable<BlockProperties> = Object.entries(blocksObj).map(
-            ([blockId, block]): BlockProperties => ({
-                id: blockId,
-                ...block,
-            }),
-        );
-        return new BlockCollection(allBlocks);
+        return new BlockCollection(makeIterable(this.prop('blocks'), deserializeBlocks));
```
Comment (mzgoddard) [inline src/sb-util.ts]: I look forward to making this configurable in the future.

File: src/sb3-serialize.ts [hunk 3]
```diff
@@ -0,0 +1,219 @@
...
+            inputs[name] = { name, block, shadow };
+        } else {
+            inputs[name] = data as BlockInput;
+        }
+    }
+
+    return { inputs, newBlocks };
+}
+
+function deserializeBlockObject(serialized: SB3SerializedBlockObject, id: string): BlockProperties[] {
```
Comment (gnarf) [inline src/sb3-serialize.ts]: oh yeah - one block "object" in sb3 land can yield multiple blocks!  how interesting!

File: tests/sb-util.test.ts [hunk 4]
```diff
@@ -1,226 +1,244 @@
-import { loadSb3, loadProjectJson, loadCloudId, ScratchProject, SpriteCollection, Sprite, BlockCollection, Block } from '../src/sb-util';
+/* eslint-disable @typescript-eslint/explicit-function-return-type */
```
Comment (gnarf) [inline tests/sb-util.test.ts]: sorry for the huge diff on this file, but you didn't have prettier running on the tests folder, so I enabled it

Comment (gnarf) [reply tests/sb-util.test.ts]: Friendly reminder this mode exists: 
image

File: src/abstracts.ts [hunk 5]
```diff
@@ -10,6 +10,70 @@ export interface SB3ProjectJSON {
...
+export enum ScratchVariableTypes {
+    SCALAR_TYPE = '',
+    LIST_TYPE = 'list',
+    BROADCAST_MESSAGE_TYPE = 'broadcast_msg',
+}
+
+export type SB3SerializedField = [any, string?];
+
+export interface SB3SerializedFields {
+    [name: string]: SB3SerializedField;
```
Comment (evmiguel) [inline src/abstracts.ts]: Does the key [name: string] mean that the key can be any string, or is it just 'name'?

Comment (gnarf) [reply src/abstracts.ts]: any string

**Pull request merged**